### PR TITLE
Moved to L7 loadbalancing demo.

### PR DESCRIPTION
**Pull request by bwplotka**
Signed-off-by: Bartlomiej Plotka <bwplotka@gmail.com>

### CODE REVIEW

File: cmd/loadbalancer/main.go [hunk 1]
```diff
@@ -60,20 +60,34 @@ func main() {
...
 		srv := &http.Server{Addr: *addr, Handler: mux}
 
+		l, err := net.Listen("tcp", *addr)
+		if err != nil {
+			log.Fatalf("new listener failed %v; exiting\n", err)
+		}
 		g.Add(func() error {
-			return srv.ListenAndServe()
+			return srv.Serve(
+				conntrack.NewInstrumentedListener(l, conntrack.NewListenerMetrics(
```
Comment (kakkoyun) [inline cmd/loadbalancer/main.go:86]: Nice!

File: pkg/lbtransport/transport_test.go [hunk 2]
```diff
@@ -0,0 +1,22 @@
+package lbtransport
+
+import (
+	"testing"
+
+	"github.com/prometheus/client_golang/prometheus/testutil"
+)
+
+func TestLBtranport(t *testing.T) {
```
Comment (kakkoyun) [inline pkg/lbtransport/transport_test.go:9]: This is out-dated, we have to update

Comment (kakkoyun) [reply pkg/lbtransport/transport_test.go:9]: I can handle it

File: pkg/lbtransport/transport.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+func NewLoadBalancingTransport(discovery Discovery, picker TargetPicker, metrics *Metrics) *Transport {
+	return &Transport{
+		discovery: discovery,
+		picker:    picker,
+		metrics:   metrics,
+		parent: &http.Transport{
+			Proxy: http.ProxyFromEnvironment,
+			DialContext: conntrack.NewInstrumentedDialContextFunc(
+				(&net.Dialer{
+					Timeout:   10 * time.Second,
```
Comment (kakkoyun) [inline pkg/lbtransport/transport.go:73]: Just to be good samaritans, we should make everything parametric and configurable.

File: pkg/lbtransport/transport.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+	dialerMetrics *conntrack.DialerMetrics
+}
+
+func NewMetrics(reg prometheus.Registerer) *Metrics {
+	var m Metrics
+
+	m.successes = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Subsystem: "lbtransport",
+		Name:      "rt_failures_total",
```
Comment (kakkoyun) [inline pkg/lbtransport/transport.go:28]: `round_tripper_failures_total` would be better and more clear

Comment (kakkoyun) [reply pkg/lbtransport/transport.go:28]: Same goes for others as well

File: pkg/conntrack/listener.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+		conn net.Conn
+		err  error
+	)
+	conn, err = ct.Listener.Accept()
+	if err != nil {
+		ct.metrics.failedAcceptsTotal.Inc()
+		return nil, err
+	}
+	if tcpConn, ok := conn.(*net.TCPConn); ok {
+		tcpConn.SetKeepAlive(true)
```
Comment (kakkoyun) [inline pkg/conntrack/listener.go:71]: Better to make them configurable/parametric rather than hardcoding.

**Comment by kakkoyun** [comment]
Awesome!

**Comment by bwplotka** [comment]
Thanks, all your comments makes sense. Can me merge for now and fix later?

File: pkg/lbtransport/transport.go (hunk 3 above)
Comment (bwplotka) [reply pkg/lbtransport/transport.go:73]: For demo it's fine. Maybe Todo comment would be enough.

**Pull request merged**